internal/controller/pkg/revision/runtime_function.go

```diff
@@ -0,0 +1,178 @@
+	}
+
+	return nil
+}
+
+// Post performs operations meant to happen after establishing objects.
+func (h *FunctionHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.PackageRevisionWithRuntime, manifests ManifestBuilder) error {
+	po, _ := xpkg.TryConvert(pkg, &pkgmetav1beta1.Function{})
+	functionMeta, ok := po.(*pkgmetav1beta1.Function)
+	if !ok {
+		return errors.New(errNotFunction)
+	}
+	if pr.GetDesiredState() != v1.PackageRevisionActive {
+		return nil
+	}
+
+	sa := manifests.ServiceAccount()
+	if err := h.client.Apply(ctx, sa); err != nil {
+		return errors.Wrap(err, errApplyFunctionSA)
+	}
+
+	d := manifests.Deployment(sa.Name, functionDeploymentOverrides(functionMeta, pr)...)
+	if err := h.client.Apply(ctx, d); err != nil {
+		return errors.Wrap(err, errApplyFunctionDeployment)
+	}
+
+	for _, c := range d.Status.Conditions {
+		if c.Type == appsv1.DeploymentAvailable {
+			if c.Status == corev1.ConditionTrue {
+				return nil
+			}
+			return errors.Errorf(errFmtUnavailableFunctionDeployment, c.Message)
+		}
+	}
+	return errors.New(errNoAvailableConditionFunctionDeployment)
+}
+
+// Deactivate performs operations meant to happen before deactivating a revision.
+func (h *FunctionHooks) Deactivate(ctx context.Context, _ v1.PackageRevisionWithRuntime, manifests ManifestBuilder) error {
+	sa := manifests.ServiceAccount()
+	// Delete the service account if it exists.
+	if err := h.client.Delete(ctx, sa); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteFunctionSA)
+	}
+
+	// Delete the deployment if it exists.
+	// Different from the Post runtimeHook, we don't need to pass the
+	// "functionDeploymentOverrides()" here, because we're only interested
+	// in the name and namespace of the deployment to delete it.
+	if err := h.client.Delete(ctx, manifests.Deployment(sa.Name)); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteFunctionDeployment)
+	}
+
+	// NOTE(ezgidemirel): Service and secret are created per package. Therefore,
+	// we're not deleting them here.
+	return nil
+}
+
+func functionDeploymentOverrides(functionMeta *pkgmetav1beta1.Function, pr v1.PackageRevisionWithRuntime) []DeploymentOverride {
+	do := []DeploymentOverride{
+		DeploymentRuntimeWithAdditionalPorts([]corev1.ContainerPort{
+			{
+				Name:          grpcPortName,
+				ContainerPort: servicePort,
+			},
+		}),
+	}
+
+	image := pr.GetSource()
+	if functionMeta.Spec.Image != nil {
+		image = *functionMeta.Spec.Image
+	}
+	do = append(do, DeploymentRuntimeWithOptionalImage(image))
+
+	return do
+}
+
+func functionServiceOverrides() []ServiceOverride {
+	return []ServiceOverride{
+		// We want a headless service so that our gRPC client (i.e. the Crossplane
+		// FunctionComposer) can load balance across the endpoints.
+		// https://kubernetes.io/docs/concepts/services-networking/service/#headless-services
+		ServiceWithClusterIP(corev1.ClusterIPNone),
```

Comment:
FWIW I had noticed recently this didn't seem to be working in master (our ClusterIP was not None) but didn't get around to investigating. Not a big deal/blocker for this PR but I'm hoping it magically fixes it.

Reply:
Just tried and can confirm that this is working fine with the PR:

```bash
❯ kubectl -n crossplane-system get svc -w
NAME                  TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
crossplane-webhooks   ClusterIP   10.96.223.5     <none>        9443/TCP   4h19m
function-dummy        ClusterIP   None            <none>        9443/TCP   7s
```